pkg/networkservice/common/heal/client.go

```diff
@@ -45,15 +57,174 @@ func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) n
+	closeOnError := false
+	u.conns.applyLockedOrNew(request.GetConnection().GetId(), func(created bool, info *connectionInfo) {
+		conn := request.GetConnection()
+		if created {
+			successVerificationCh = make(chan struct{}, 1)
+
+			info.conn = conn.Clone()
+			info.state = connectionstateAwaitingConfirmation
+			info.successVerificationCh = successVerificationCh
+			closeOnError = true
+		} else if conn.Path != nil && int(conn.Path.Index) < len(conn.Path.PathSegments)-1 {
+			storedConn := info.conn
+			path := request.GetConnection().Path
+			path.PathSegments = path.PathSegments[:path.Index+1]
+			path.PathSegments = append(path.PathSegments, storedConn.Path.PathSegments[path.Index+1:]...)
+			conn.NetworkServiceEndpointName = storedConn.NetworkServiceEndpointName
+			if info.state == connectionstateBroken {
+				successVerificationCh = make(chan struct{}, 1)
+				info.successVerificationCh = successVerificationCh
+			}
+		}
+	})
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+	if successVerificationCh != nil {
+		err = u.confirmConnectionSuccess(conn, successVerificationCh)
+		if err != nil {
+			if closeOnError {
+				u.removeConnectionFromMonitor(conn)
+				_, _ = next.Client(ctx).Close(ctx, request.GetConnection(), opts...)
+			}
+			return nil, err
+		}
 	}
-	return conn, err
+	u.conns.applyLocked(conn.GetId(), func(info *connectionInfo) {
+		info.conn = conn.Clone()
+	})
+
+	return conn, nil
 }
 
 func (u *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	if cachedConn, ok := u.removeConnectionFromMonitor(conn); ok {
+		conn = cachedConn
+	} else {
+		return &empty.Empty{}, nil
+	}
+
 	return next.Client(ctx).Close(ctx, conn, opts...)
 }
+
+// listenToConnectionChanges - loops on events from MonitorConnectionClient while the monitor client is alive.
+//                             Updates connection cache and broadcasts events of successful connections.
+//                             Calls heal when something breaks.
+func (u *healClient) listenToConnectionChanges(heal requestHealFuncType, currentPathSegment *networkservice.PathSegment, errCh chan error) {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{
+		PathSegments: []*networkservice.PathSegment{{Name: currentPathSegment.Name}, {Name: ""}},
+	})
+	if err != nil {
+		errCh <- errors.Wrap(err, "MonitorConnections failed")
+		return
+	}
+
+	close(errCh)
+
+	for {
+		event, err := monitorClient.Recv()
+		if err != nil {
+			u.conns.Range(func(id string, info *connectionInfo) bool {
+				info.mut.Lock()
+				defer info.mut.Unlock()
+				heal(info.conn, true)
+				return true
+			})
+			return
+		}
+
+		for _, eventConn := range event.GetConnections() {
+			eventConn := eventConn
```

Comment:
```suggestion
```

Reply:
Without this line linter complains that I have unsafe usage of loop variable.

Reply:
OK, I see, linter is correct.

Then let's rework this to the index loop.

```suggestion
			eventConn := event.GetConnections()[i]
```